hbase-server/src/main/java/org/apache/hadoop/hbase/master/ServerManager.java

```diff
@@ -718,14 +720,19 @@ public static void closeRegionSilentlyAndWait(ClusterConnection connection, Serv
   public AdminService.BlockingInterface getRsAdmin(final ServerName sn) throws IOException {
     AdminService.BlockingInterface admin = this.rsAdmins.get(sn);
     if (admin == null) {
-      LOG.debug("New admin connection to " + sn.toString());
-      if (sn.equals(master.getServerName()) && master instanceof HRegionServer) {
-        // A master is also a region server now, see HBASE-10569 for details
-        admin = ((HRegionServer) master).getRSRpcServices();
-      } else {
-        admin = this.connection.getAdmin(sn);
-      }
-      this.rsAdmins.put(sn, admin);
+      return this.rsAdmins.computeIfAbsent(sn, server -> {
+        LOG.debug("New admin connection to " + server.toString());
+        if (server.equals(master.getServerName()) && master instanceof HRegionServer) {
+          // A master is also a region server now, see HBASE-10569 for details
+          return ((HRegionServer) master).getRSRpcServices();
+        } else {
+          try {
+            return this.connection.getAdmin(server);
+          } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
We should throw this IOException out? This is a problem of computeIfAbsent, as the functional interface does not throw IOException...

Reply:
Yeah, good point

Reply:
`computeIfAbsentEx` seems like a good idea here. I don't have an informed opinion regarding whether the caching is really necessary, but I'm happy to rip it out if we think that's worth doing now

Reply:
Ah, checked the code, as @bbeaudreault mentioned, we already have a cache in `ConnectionImplementation`, https://github.com/apache/hbase/blob/18a38aeec6b6db4abf24936bc8a403bca58e7e32/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionImplementation.java#L1434

So I think we can just remove the cache here.